src/cgame/cg_rocket_draw.cpp

```diff
@@ -2174,6 +2174,130 @@ class PredictedMineEfficiencyElement : public HudElement
 	std::unordered_map<int, std::string> pluralSuffix;
 };
 
+class BarbsHudElement : public HudElement
+{
+public:
+	BarbsHudElement ( const Rocket::Core::String& tag ) :
+	HudElement ( tag, ELEMENT_ALIENS ),
+	numBarbs( 0 ),
+	maxBarbs( BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo ),
+	regenerationInterval ( 0 ),
+	t0 ( 0 ),
+	offset ( 0 ) {}
+
+	void OnAttributeChange( const Rocket::Core::AttributeNameList& changed_attributes )
+	{
+		HudElement::OnAttributeChange( changed_attributes );
+		if ( changed_attributes.find( "src" ) != changed_attributes.end() )
+		{
+			if ( maxBarbs > 0 )
+			{
+				Rocket::Core::String src = GetAttribute<Rocket::Core::String>( "src", "" );
+				Rocket::Core::String base( va("<img class='barbs' src='%s' />", src.CString() ) );
+				Rocket::Core::String rml;
+
+				for ( int i = 0; i < maxBarbs; i++ )
+				{
+					rml += base;
+				}
+				SetInnerRML( rml );
+			}
+			else
+			{
+				SetInnerRML( "" );
+			}
+		}
+	}
+
+	void DoOnUpdate()
+	{
+		int newNumBarbs = cg.snap->ps.ammo;
+		int interval = BG_GetBarbRegenerationInterval( cg.snap->ps );
+
+		if ( newNumBarbs < maxBarbs )
+		{
+			// start regenerating barb now
+			if ( newNumBarbs > numBarbs || ( newNumBarbs < numBarbs && numBarbs == maxBarbs ) )
+			{
+				t0 = cg.time;
+				// sin(-pi/2) is minimal
+				offset = -M_PI_2;
+				assert( GetSin() == -1.0 );
+			}
+			// change regeneration speed
+			else if ( interval != regenerationInterval )
+			{
+				float sOld = GetSin();
+				float cOld = GetCos();
+
+				// avoid sudden jumps in opacity
+				t0 = cg.time;
+				if ( cOld >= 0.0 )
+				{
+					offset = asin( sOld );
+				}
+				else
+				{
+					offset = M_PI - asin( sOld );
+				}
+				regenerationInterval = interval;
+
+				assert( fabs(GetSin() - sOld ) < 0.001);
+				assert( GetCos() * cOld > 0 );
+			}
+		}
+		numBarbs = newNumBarbs;
+
+		for ( int i = 0; i < GetNumChildren(); i++ )
```

Comment:
I'll try some things after we merge. I think displaying missing barbs at 25% and existing barbs at 100% opacity and pulsating between 25% and 50% could work. For now this looks good.